### PR TITLE
Make i128 aligned to 16 bytes on all 64bit plats

### PR DESCRIPTION
**Pull request by nagisa**
Fixes https://github.com/rust-lang/rust/issues/38762

This is based on https://github.com/rust-lang/rust/pull/38870.

### CODE REVIEW

**Comment by eddyb** [comment]
Why all architectures? I only heard something about x64 SysV. Do you have links to more information?

File: src/librustc_trans/cabi_x86_64.rs [hunk 1]
```diff
@@ -343,6 +343,7 @@ pub fn compute_abi_info(ccx: &CrateContext, fty: &mut FnType) {
                 arg.make_indirect(ccx);
                 if let Some(attr) = ind_attr {
                     arg.attrs.set(attr);
+                    arg.attrs.set_align(ty_align(arg.ty, 0) as u64);
```
Comment (eddyb) [inline src/librustc_trans/cabi_x86_64.rs]: Isn't this a noop? (same for all the other `cabi_*` changes)

Comment (nagisa) [inline src/librustc_trans/cabi_x86_64.rs]: See https://github.com/rust-lang/rust/pull/38870 for this.

**Comment by nagisa** [comment]
> Why all architectures? I only heard something about x64 SysV. Do you have links to more information?

Only 64 bit ones¹. Because I have no idea what expected alignment on those arches are for i128. Having larger alignment is not incorrect, only potentially wasteful, whereas alignment being 8 instead of expected 16 might end up in similar hard-to-debug issues again.

I’d rather have all 64 bit arches have i128 aligned to 16 bytes and reduce the alignment on-demand, rather than go around looking for some weird “codegen bug” where something somewhere gets overwritten on stack on some very obscure platform.

¹: 32 bit arches do not matter because C compilers do not support int128 on 32 bit machines.

**Comment by eddyb** [comment]
> Having larger alignment is not incorrect, only potentially wasteful

... except when using inside a `struct` where you have alignment padding.

If you found clang code giving custom alignments to `i128`, please link to it, that'd be a better starting point.

**Comment by nagisa** [comment]
> If you found clang code giving custom alignments to i128

That’s https://github.com/llvm-mirror/clang/blob/ec16cb9b5a481d62a73ad47fa59034ced4d62022/lib/AST/ASTContext.cpp#L416, also linked in the referenced issue.

**Comment by eddyb** [comment]
> also linked in the referenced issue

Oops, had not read that in detail.

**Comment by eddyb** [comment]
clang has `// int128_t is 128-bit aligned on all targets.` which agrees with this PR, should be in description.
Although I'm wondering if at this point patching LLVM is not a better solution, seems like their bug.

**Comment by nagisa** [comment]
Better solved in LLVM.

**Pull request closed without merging**